test/IFrame_test.js

```diff
@@ -0,0 +1,719 @@
+      assert.equal(player.playbackRate, 1.5);
+      assert(autoUpdateTriggered);
+      autoUpdateTriggered = false;
+    });
+
+    it('should continue playing on playing seek', function() {
+      doc.changeFirestore(40, 0.5, true);
+      getRealtimeUpdates();
+      onPlayerStateChange(); // play state change
+      assert.equal(player.timestamp, 40);
+      assert.equal(player.playerState, 1);
+      assert.equal(player.playbackRate, 0.5);
+      assert(!autoUpdateTriggered);
+    });
+  });
+
+  describe('Update on Pause', function() {
+    it('should pause when playing', function() {
+      doc.changeFirestore(10, 1, false);
+      getRealtimeUpdates();
+      onPlayerStateChange(); // pause state change
+      assert.equal(player.timestamp, 10);
+      assert.equal(player.playerState, 2);
+      assert.equal(player.playbackRate, 1);
+      assert(!autoUpdateTriggered);
+    });
+
+    it('should remain paused on seek while paused', function() {
+      doc.changeFirestore(30, 1, false);
+      getRealtimeUpdates();
+      onPlayerStateChange(); // pause state change
+      assert.equal(player.timestamp, 30);
+      assert.equal(player.playerState, 2);
+      assert.equal(player.playbackRate, 1);
+      assert(!autoUpdateTriggered);
+    });
+  });
+
+  describe('Playing Seek Boundaries', function() {
+    it('should not seek for less than 5 second time difference' +
+        'when playing', function() {
+      doc.changeFirestore(31, 1, true);
+      getRealtimeUpdates();
+      onPlayerStateChange();
+      assert.equal(player.timestamp, 30);
+    });
+
+    it('should not seek for 5 second time difference when playing',
+        function() {
+          doc.changeFirestore(35, 1, true);
+          getRealtimeUpdates();
+          onPlayerStateChange();
+          assert.equal(player.timestamp, 30);
+        });
+
+    it('should halve sync window when playing at half speed',
+        function() {
+          doc.changeFirestore(33, 0.5, true); // add only 3 seconds (> 2.5)
+          getRealtimeUpdates();
+          onPlayerStateChange();
+          assert.equal(player.timestamp, 33);
+          // still updates because at halfspeed
+          // 5 realtime seconds is 2.5 video seconds
+        });
+
+    it('should double sync widnow when playing at double speed',
+        function() {
+          doc.changeFirestore(40, 2, true); // add 7 seconds (< 10 seconds)
+          getRealtimeUpdates();
+          onPlayerStateChange();
+          assert.equal(player.timestamp, 33);
+          autoUpdateTriggered = false;
+        });
+    // Note that when I say double or halve sync window, I mean solely
+    // in the case of deciding whether or not to move the user's playhead.
+    // Playback speed has no effect on other uses for sync window (i.e.
+    // how often auto updates are sent).
+  });
+});
+/* eslint-enable no-undef*/
+
+// ATTENTION!!!
+
+// Below is modified version of code to work with mocks
```

Comment:
The reason for writing the tests is to make sure if the logic implemented works as intended. On first glance, it seems like you implemented methods that would mimic the main logic for mocking which doesn't seem right.

Any reason you can't use the main logic?

Reply:
Most of the code I deleted made calls to Firestore instead of the mocks. For example, in the original code, I used:

// queueDataRef.orderBy('requestTime', 'asc').limit(1).get()
      // .then(function(querySnapshot) {
        // querySnapshot.forEach(function(doc) {

to get an object with information on the next video in the queue's ID and thumbnail. It seemed unnecessary to make a mock which has all these methods and works as a promise, so I called it with one method in the mock version.
    Another problem happened where if I was testing a part of the code which sets a timeout, its effects might happen while another set timeout code was running and they would both be changing my mock Firestore at the same time, so I made the mock version call things more immediately where I could. An example of something like this is the "autoUpdateTriggered" variable where instead of setting a timeout to calling updateInfo again and again, which would send info to my mock, it just sets a variable to say it would have looped sending this information to Firestore. 
    Also, when I first started, it seemed like the code couldn't access some of the html, so I made stand in variables which played the roles of certain tags' display attributes. Essentially, mimicking the logic made the mocks and testing easier than deciding how to deal with issues with timing, automatic listeners, access to html, and calls to Firestore. I was unaware that this would be a problem, but I understand that using the unaltered code is more helpful. How should I go about addressing some of these issues?